src/libtorch.cc

```diff
@@ -253,8 +255,28 @@ ModelState::LoadModel(
 
   try {
     std::istringstream model_stream(model_data_str);
-    torch_model->reset(
-        new torch::jit::Module(torch::jit::load(model_stream, device)));
+    if (kind == TRITONSERVER_INSTANCEGROUPKIND_MODEL) {
+      torch_model->reset(
+          new torch::jit::Module(torch::jit::load(model_stream)));
+
+      // Get the device used in the model
+      auto parameters = (*torch_model)->parameters();
```

Comment:
There is also the `attributes` field: https://pytorch.org/cppdocs/api/structtorch_1_1jit_1_1_module.html#_CPPv4NK5torch3jit6Module10attributesEb

Do we need to include that in the list hear too?

Reply:
Nice catch! Missed the `attributes` field.
However, there seems to be no proper way to retrieve all the device used in a model based on the [comment](https://github.com/pytorch/pytorch/issues/7460#issuecomment-388013922). Although we can get the attributes, parameters and the buffers from a model, it is possible that the model contains complex structures or custom-defined objects where the device information may not be directly accessible. Additionally, it's challenging to test every corner case as we may not have access to all possible models. If a device id is missed, it can lead to data corruption because the corresponding CUDA stream is not synchronized. To address this, I was thinking we could create a new CUDA stream for every available device. This ensures that each stream is properly synchronized. Let me know your thoughts on this.